joomla/joomla-cms
# [4.0] Add upload file size check to Joomla Update Component Upload & Update

Pull Request for Issue #27494 .

### Summary of Changes

This Pull Request (PR) adds following to the Upload & Update tab of the Joomla Update Component:
- Display of the selected upload file's size.
- Check if the selected file is too big to be uploaded with respect to the PHP limits.

If the selected upload file is too big, a warning alert is shown. If the user tries to upload that file anway, the upload is rejected with an appropriate browser alert message.

The size of the selected file is always shown in MB because the PHP limits and so the maximum file size will with a 99.9 % likelyhood also be in MB.

For the Upload & Install Package tab of the Extension Installer I've created PR #27665 with similar changes, but there only the browser alert is implemented. Display of the size of the selected files is not really possible there.

### Testing Instructions

**Requirements:** npm

1. Apply patch of this PR.
2. Run `npm run build:js` (or `npm install` if you prefer this).
3. Clear browser cache to get rid of old cached JS.
4. Go to `administrator/index.php?option=com_joomlaupdate` and there to tab "Upload & Update".
5. Check the `Joomla package file` fields.
**Result:** See section "Expected result" screenshot 1. No information related to the selected file's size is shown if no file is selected.
6. Select a file which is **not** too big compared to the maximum upload size shown below the file selection.
**Hint:** This can be any file, it doesn't necessarily have to be a Joomla Update package.
**Result:** See section "Expected result" screenshot 2. The size of the selected file is shown below the maximum upload size.
7. Select a file which **is** too big compared to the maximum upload size shown below the file selection.
**Hint:** This can be any file, it doesn't necessarily have to be a Joomla Update package.
**Result:** See section "Expected result" screenshot 3. The size of the selected file is shown below the maximum upload size, and below that a warning alert is shown telling that the selected file is too big.
8. Try to upload the file anway by selecting the `Upload & Install` button.
**Result:** See section "Expected result" screenshot 4. A browser alert is shown telling that the file can't be uploaded because it's too big. The upload is not started.
9. Toggle the file selection a few times between too big and not too big file, and also between different files of each type.
**Result:** The warning disappears when the selection is changed from too big to not too big, and it appears again when vice versa. The file size is updated with each change of the file.
10. Finally select a valid Joomla Update package which is not too big and verify that uploading still works.
**Result:** Uploading of not too big file still works as well as without this PR.

### Expected result

- No file selected:
![snap-01-no-file-selected](https://user-images.githubusercontent.com/7413183/72682643-97f0bf80-3acf-11ea-9fd1-399ab349e381.png)
- File size ok:
![snap-02-file-ok](https://user-images.githubusercontent.com/7413183/73138300-7f4d5000-4061-11ea-9f50-9bd500a3c6d3.png)
- File too big:
![snap-03-file-too-big](https://user-images.githubusercontent.com/7413183/73138306-85dbc780-4061-11ea-8910-edf08292e536.png)
- Broswer alert when trying to upload too big file:
![snap-04-file-too-big-upload](https://user-images.githubusercontent.com/7413183/73138312-8c6a3f00-4061-11ea-8054-48d4f9945c16.png)

### Actual result

See isue #27494 
- If `post_max_size` is big enough but `upload_max_size` is violated:
![j4_upload-max-filesize-violated_com-joomlaupdate](https://user-images.githubusercontent.com/7413183/72662695-50dcce80-39ea-11ea-99d3-49c0b1095e33.png)
**Hint:** The 403 and the missing language strings are separate issues not being fixed with this PR here. The 403 could also be related to a problem in my testing environment.

- If `upload_max_size` is big enough but `post_max_size` is violated:
Nothing happens, and PHP error log shows: `PHP Warning: POST Content-Length of 22972015 bytes exceeds the limit of 16777216 bytes in Unknown on line 0, referer: ...` (with byte figures suitable to your test).

### Documentation Changes Required

Nothing special for this PR. The help screens for J4 backend have to be udpated anyway.
